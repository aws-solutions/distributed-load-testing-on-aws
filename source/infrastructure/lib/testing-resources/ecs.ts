// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnResource } from "aws-cdk-lib";
import { CfnCluster, CfnTaskDefinition } from "aws-cdk-lib/aws-ecs";
import {
  Effect,
  ManagedPolicy,
  ServicePrincipal,
  Role,
  PolicyDocument,
  PolicyStatement,
  Policy,
} from "aws-cdk-lib/aws-iam";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { CfnSecurityGroup, CfnSecurityGroupEgress, CfnSecurityGroupIngress } from "aws-cdk-lib/aws-ec2";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export interface ECSResourcesConstructProps {
  readonly cloudWatchLogsPolicy: Policy;
  // Container image
  readonly containerImage: string;
  // Fargate VPC ID
  readonly fargateVpcId: string;
  // Scenarios S3 bucket
  readonly scenariosS3Bucket: string;
  // IP CIDR for Fargate egress
  readonly securityGroupEgress: string;
  // Solution ID
  readonly solutionId: string;
}

/**
 * Distributed Load Testing on AWS Fargate and ECS test runner construct.
 * This creates the ECS cluster, Fargate task definition, and Security Group
 */
export class ECSResourcesConstruct extends Construct {
  public taskClusterName: string;
  public ecsCloudWatchLogGroup: LogGroup;
  public taskDefinitionArn: string;
  public taskExecutionRoleArn: string;
  public ecsSecurityGroupId: string;
  public taskDefinitionFamily: string | undefined;

  constructor(scope: Construct, id: string, props: ECSResourcesConstructProps) {
    super(scope, id);

    const dltTaskCluster = new CfnCluster(this, "DLTEcsCluster", {
      clusterName: Aws.STACK_NAME,
      clusterSettings: [{ name: "containerInsights", value: "enabled" }],
      tags: [
        {
          key: "SolutionId",
          value: props.solutionId,
        },
        {
          key: "CloudFormation Stack",
          value: Aws.STACK_NAME,
        },
      ],
    });

    this.taskClusterName = dltTaskCluster.ref;

    const scenariosBucketArn = Bucket.fromBucketName(this, "ScenariosBucket", props.scenariosS3Bucket).bucketArn;

    const dltTaskExecutionRole = new Role(this, "DLTTaskExecutionRole", {
      assumedBy: new ServicePrincipal("ecs-tasks.amazonaws.com"),
      managedPolicies: [ManagedPolicy.fromAwsManagedPolicyName("service-role/AmazonECSTaskExecutionRolePolicy")],
      inlinePolicies: {
        ScenariosS3Policy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:HeadObject", "s3:PutObject", "s3:GetObject", "s3:ListBucket"],
              resources: [scenariosBucketArn, `${scenariosBucketArn}/*`],
            }),
          ],
        }),
      },
    });
    dltTaskExecutionRole.attachInlinePolicy(props.cloudWatchLogsPolicy);
    this.taskExecutionRoleArn = dltTaskExecutionRole.roleArn;

    this.ecsCloudWatchLogGroup = new LogGroup(this, "DLTCloudWatchLogsGroup", {
      retention: RetentionDays.ONE_YEAR,
    });
    const dltLogsGroupResource = this.ecsCloudWatchLogGroup.node.defaultChild as CfnResource;
    dltLogsGroupResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W84",
          reason: "KMS encryption unnecessary for log group",
        },
      ],
    });

    const dltTaskDefinition = new CfnTaskDefinition(this, "DLTTaskDefinition", {
      cpu: "2048",
      memory: "4096",
      networkMode: "awsvpc",
      executionRoleArn: this.taskExecutionRoleArn,
      requiresCompatibilities: ["FARGATE"],
      taskRoleArn: this.taskExecutionRoleArn,
      family: `${Aws.STACK_NAME}-task-family`,
      containerDefinitions: [
        {
          essential: true,
          name: `${Aws.STACK_NAME}-load-tester`,
          image: props.containerImage,
          memory: 4096,
          logConfiguration: {
            logDriver: "awslogs",
            options: {
              "awslogs-group": this.ecsCloudWatchLogGroup.logGroupName,
              "awslogs-stream-prefix": "load-testing",
              "awslogs-region": `${Aws.REGION}`,
            },
          },
        },
      ],
    });

    this.taskDefinitionArn = dltTaskDefinition.ref;
    this.taskDefinitionFamily = dltTaskDefinition.family;

    const ecsSecurityGroup = new CfnSecurityGroup(this, "DLTEcsSecurityGroup", {
      vpcId: props.fargateVpcId,
      groupDescription: "DLTS Tasks Security Group",
    });
    ecsSecurityGroup.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W40",
          reason: "IpProtocol set to -1 (any) as ports are not known prior to running tests",
        },
      ],
    });

    this.ecsSecurityGroupId = ecsSecurityGroup.ref;

    new CfnSecurityGroupEgress(this, "DLTSecGroupEgress", {
      cidrIp: props.securityGroupEgress,
      description: "Allow tasks to call out to external resources",
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "-1",
    });

    new CfnSecurityGroupIngress(this, "DLTSecGroupIngress", {
      description: "Allow tasks to communicate",
      fromPort: 50000,
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "tcp",
      sourceSecurityGroupId: ecsSecurityGroup.ref,
      toPort: 50000,
    });
  }
}
