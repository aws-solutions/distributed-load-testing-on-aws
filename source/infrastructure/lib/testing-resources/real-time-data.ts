// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, CfnResource, Duration, Stack } from "aws-cdk-lib";
import { Code, Function as LambdaFunction, Runtime } from "aws-cdk-lib/aws-lambda";
import { Effect, PolicyStatement, PolicyDocument, Role, ServicePrincipal, Policy } from "aws-cdk-lib/aws-iam";
import { IBucket } from "aws-cdk-lib/aws-s3";
import { LogGroup, FilterPattern } from "aws-cdk-lib/aws-logs";
import { LambdaDestination } from "aws-cdk-lib/aws-logs-destinations";
import { Construct } from "constructs";

export interface RealTimeDataConstructProps {
  readonly cloudWatchLogsPolicy: Policy;
  readonly ecsCloudWatchLogGroup: LogGroup;
  readonly iotEndpoint: string;
  readonly mainRegion: string;
  /**
   * Solution config properties.
   * solution ID, version, source code bucket, and source code prefix
   */
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly sourceCodeBucket: IBucket;
  readonly sourceCodePrefix: string;
}

export class RealTimeDataConstruct extends Construct {
  constructor(scope: Construct, id: string, props: RealTimeDataConstructProps) {
    super(scope, id);

    const realTimeDataPublisherRole = new Role(this, "realTimeDataPublisherRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        IoTPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Publish"],
              resources: [
                Stack.of(this).formatArn({
                  region: props.mainRegion,
                  service: "iot",
                  resource: "topic",
                  resourceName: "*",
                  arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
                }),
              ],
            }),
          ],
        }),
      },
    });
    realTimeDataPublisherRole.attachInlinePolicy(props.cloudWatchLogsPolicy);

    const realTimeDataPublisher = new LambdaFunction(this, "RealTimeDataPublisher", {
      description: "Real time data publisher",
      handler: "index.handler",
      role: realTimeDataPublisherRole,
      code: Code.fromBucket(props.sourceCodeBucket, `${props.sourceCodePrefix}/real-time-data-publisher.zip`),
      runtime: Runtime.NODEJS_18_X,
      timeout: Duration.seconds(180),
      environment: {
        MAIN_REGION: props.mainRegion,
        IOT_ENDPOINT: props.iotEndpoint,
        SOLUTION_ID: props.solutionId,
        VERSION: props.solutionVersion,
      },
    });

    const realTimeDataPublisherResource = realTimeDataPublisher.node.defaultChild as CfnResource;
    realTimeDataPublisherResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "This Lambda function does not require a VPC",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });

    const ecsCloudWatchSubscriptionFilter = props.ecsCloudWatchLogGroup.addSubscriptionFilter(
      "ECSLogSubscriptionFilter",
      {
        destination: new LambdaDestination(realTimeDataPublisher),
        filterPattern: FilterPattern.allTerms("INFO: Current:", "live=true"),
      }
    );
    const subscriptionFilterPermission = ecsCloudWatchSubscriptionFilter.node.findChild("CanInvokeLambda");
    if (subscriptionFilterPermission != null)
      ecsCloudWatchSubscriptionFilter.node.addDependency(subscriptionFilterPermission);
  }
}
