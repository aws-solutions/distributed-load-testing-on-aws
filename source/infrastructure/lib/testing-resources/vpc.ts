// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Tags } from "aws-cdk-lib";
import { IpAddresses, SubnetType, Vpc } from "aws-cdk-lib/aws-ec2";
import { Construct } from "constructs";

/**
 * Distributed Load Testing on AWS VPC construct.
 * Creates a VPC for the Fargate test tasks.
 * Includes 2 public subnets across 2 availability zones.
 */
export class FargateVpcConstruct extends Construct {
  public vpc: Vpc;

  constructor(scope: Construct, id: string, vpcCidrBlock: string) {
    super(scope, id);

    const fargateVpc = new Vpc(this, "DLTFargateVpc", {
      ipAddresses: IpAddresses.cidr("192.168.0.0/16"),
      enableDnsHostnames: true,
      enableDnsSupport: true,
      subnetConfiguration: [
        {
          name: "DLTSubnetA",
          subnetType: SubnetType.PUBLIC,
          cidrMask: 20,
        },
        {
          name: "DLTSubnetB",
          subnetType: SubnetType.PUBLIC,
          cidrMask: 20,
        },
      ],
      maxAzs: 2,
      natGateways: 0,
    });

    fargateVpc.node.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W60",
          reason: "This VPC is used for the test runner Fargate tasks only, it does not require VPC flow logs.",
        },
      ],
    });

    this.vpc = fargateVpc;

    Tags.of(fargateVpc).add("Name", Aws.STACK_NAME);
  }
}
