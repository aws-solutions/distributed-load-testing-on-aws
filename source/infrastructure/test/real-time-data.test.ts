// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Template } from "aws-cdk-lib/assertions";
import { App, DefaultStackSynthesizer, Stack } from "aws-cdk-lib";
import { Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { LogGroup } from "aws-cdk-lib/aws-logs";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { RealTimeDataConstruct } from "../lib/testing-resources/real-time-data";

test("DLT real time data resources Test", () => {
  const app = new App();
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
    }),
  });
  const testPolicy = new Policy(stack, "TestPolicy", {
    statements: [
      new PolicyStatement({
        resources: ["*"],
        actions: ["cloudwatch:Get*"],
      }),
    ],
  });
  const testLogGroup = new LogGroup(stack, "TestLogsGroup");
  const testBucket = Bucket.fromBucketName(stack, "SourceCodeBucket", "testbucket");

  const realTimeData = new RealTimeDataConstruct(stack, "TestECS", {
    cloudWatchLogsPolicy: testPolicy,
    ecsCloudWatchLogGroup: testLogGroup,
    iotEndpoint: "iotEndpoint",
    mainRegion: "test-region-1",
    solutionId: "testID",
    solutionVersion: "testVersion",
    sourceCodeBucket: testBucket,
    sourceCodePrefix: "testPrefix",
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
  Template.fromStack(stack).hasResourceProperties("AWS::Lambda::Function", {
    Code: {
      S3Bucket: "testbucket",
      S3Key: "testPrefix/real-time-data-publisher.zip",
    },
    Environment: {
      Variables: {
        MAIN_REGION: "test-region-1",
        IOT_ENDPOINT: "iotEndpoint",
        SOLUTION_ID: "testID",
        VERSION: "testVersion",
      },
    },
    Handler: "index.handler",
    Runtime: "nodejs18.x",
    Timeout: 180,
  });
  Template.fromStack(stack).hasResourceProperties("AWS::Logs::SubscriptionFilter", {
    FilterPattern: '"INFO: Current:" "live=true"',
  });
  Template.fromStack(stack).hasResourceProperties("AWS::IAM::Role", {
    Policies: [
      {
        PolicyDocument: {
          Statement: [
            {
              Action: "iot:Publish",
              Effect: "Allow",
              Resource: {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      Ref: "AWS::Partition",
                    },
                    ":iot:test-region-1:",
                    {
                      Ref: "AWS::AccountId",
                    },
                    ":topic/*",
                  ],
                ],
              },
            },
          ],
        },
      },
    ],
  });
});
