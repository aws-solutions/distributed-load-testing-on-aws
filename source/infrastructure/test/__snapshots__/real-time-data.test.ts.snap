// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT real time data resources Test 1`] = `
{
  "Resources": {
    "TestECSRealTimeDataPublisherF2114D5B": {
      "DependsOn": [
        "TestECSrealTimeDataPublisherRoleEFE9F1CD",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            {
              "id": "W89",
              "reason": "This Lambda function does not require a VPC",
            },
            {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": "testbucket",
          "S3Key": "testPrefix/real-time-data-publisher.zip",
        },
        "Description": "Real time data publisher",
        "Environment": {
          "Variables": {
            "IOT_ENDPOINT": "iotEndpoint",
            "MAIN_REGION": "test-region-1",
            "SOLUTION_ID": "testID",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestECSrealTimeDataPublisherRoleEFE9F1CD",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 180,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestECSrealTimeDataPublisherRoleEFE9F1CD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "iot:Publish",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:test-region-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":topic/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "IoTPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestLogsGroup54B681C7": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestLogsGroupECSLogSubscriptionFilterBFBFAB24": {
      "DependsOn": [
        "TestLogsGroupECSLogSubscriptionFilterCanInvokeLambdaD7381D91",
      ],
      "Properties": {
        "DestinationArn": {
          "Fn::GetAtt": [
            "TestECSRealTimeDataPublisherF2114D5B",
            "Arn",
          ],
        },
        "FilterPattern": ""INFO: Current:" "live=true"",
        "LogGroupName": {
          "Ref": "TestLogsGroup54B681C7",
        },
      },
      "Type": "AWS::Logs::SubscriptionFilter",
    },
    "TestLogsGroupECSLogSubscriptionFilterCanInvokeLambdaD7381D91": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "TestECSRealTimeDataPublisherF2114D5B",
            "Arn",
          ],
        },
        "Principal": "logs.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "TestLogsGroup54B681C7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestPolicyCC05E598": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": [
          {
            "Ref": "TestECSrealTimeDataPublisherRoleEFE9F1CD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
